// webdriverio/build/scripts/isElementClickable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = isElementClickable;

function isElementClickable(elem) {
  if (!elem.getBoundingClientRect || !elem.scrollIntoView || !elem.contains || !elem.getClientRects || !document.elementFromPoint) {
    return false;
  }

  function getOverlappingElement(elem) {
    const elemDimension = elem.getBoundingClientRect();
    const x = elemDimension.left + elem.clientWidth / 2;
    const y = elemDimension.top + elem.clientHeight / 2;
    return document.elementFromPoint(x, y);
  }

  function getOverlappingRects(elem) {
    const elems = [];
    const rects = elem.getClientRects();
    const rect = rects[0];
    const x = rect.left + rect.width / 2;
    const y = rect.top + rect.height / 2;
    elems.push(document.elementFromPoint(x, y));
    return elems;
  }

  function getOverlappingElements(elem) {
    return [getOverlappingElement(elem), ...getOverlappingRects(elem)];
  }

  function isOverlappingElementMatch(elementsFromPoint, elem) {
    return elementsFromPoint.some(elementFromPoint => elementFromPoint === elem || elem.contains(elementFromPoint));
  }

  function isElementInViewport(elem) {
    if (!elem.getBoundingClientRect) {
      return false;
    }

    const rect = elem.getBoundingClientRect();
    const windowHeight = window.innerHeight || document.documentElement.clientHeight;
    const windowWidth = window.innerWidth || document.documentElement.clientWidth;
    const vertInView = rect.top <= windowHeight && rect.top + rect.height > 0;
    const horInView = rect.left <= windowWidth && rect.left + rect.width > 0;
    return vertInView && horInView;
  }

  function isClickable(elem) {
    return isElementInViewport(elem) && elem.disabled !== true && isOverlappingElementMatch(getOverlappingElements(elem), elem);
  }

  if (!isClickable(elem)) {
    elem.scrollIntoView({
      block: 'nearest',
      inline: 'nearest'
    });

    if (!isClickable(elem)) {
      elem.scrollIntoView({
        block: 'center',
        inline: 'center'
      });
      return isClickable(elem);
    }
  }

  return true;
}
